File: lab2/cpp/galois_field_test.cpp
#include <iostream>
#include <gtest/gtest.h>

#include "GaloisField.hpp"

constexpr uint64_t PRIME = 5;

TEST(GaloisField, CreatingCorrectObjects) {
    ASSERT_EQ(GaloisField<PRIME>{}.getCharacteristic(), PRIME);
    
    GaloisField<PRIME> gf1;
    ASSERT_EQ(gf1, 0);

    GaloisField<PRIME> gf2{3};
    ASSERT_EQ(gf2, 3);

    GaloisField<PRIME> gf3{7};
    ASSERT_EQ(gf3, 2);

    GaloisField<PRIME> gf4{-16};
    ASSERT_EQ(gf4, 4);
}

TEST(GaloisField, Addition) {
    GaloisField<PRIME> gf1{3};
    GaloisField<PRIME> gf2{4};

    ASSERT_EQ(gf1 + gf2, 2);
    ASSERT_EQ(gf1 + 2, 0);
    ASSERT_EQ(2 + gf1, 0);

    gf1 += 15;
    ASSERT_EQ(gf1, 3);

    gf1 += gf2;
    ASSERT_EQ(gf1, 2);
}

TEST(GaloisField, Subtraction) {
    GaloisField<PRIME> gf1{3};
    GaloisField<PRIME> gf2{4};

    ASSERT_EQ(gf1 - gf2, 4);
    ASSERT_EQ(gf1 - 2, 1);
    ASSERT_EQ(2 - gf1, 4);

    gf1 -= 15;
    ASSERT_EQ(gf1, 3);

    gf1 -= gf2;
    ASSERT_EQ(gf1, 4);
}

TEST(GaloisField, Multiplication) {
    GaloisField<PRIME> gf1{3};
    GaloisField<PRIME> gf2{4};

    ASSERT_EQ(gf1 * gf2, 2);
    ASSERT_EQ(gf1 * 2, 1);
    ASSERT_EQ(2 * gf1, 1);

    gf1 *= 6;
    ASSERT_EQ(gf1, 3);

    gf1 *= gf2;
    ASSERT_EQ(gf1, 2);
}

TEST(GaloisField, Division) {
    GaloisField<PRIME> gf1{3};
    GaloisField<PRIME> gf2{4};

    ASSERT_EQ(gf1 / gf2, 2);
    ASSERT_EQ(gf1 / 2, 4);
    ASSERT_EQ(2 / gf1, 4);

    gf1 /= 3;
    ASSERT_EQ(gf1, 1);

    gf1 /= gf2;
    ASSERT_EQ(gf1, 4);
}

TEST(GaloisField, ZeroDivision) {
    GaloisField<PRIME> gf1{3};
    GaloisField<PRIME> gf2{0};

    EXPECT_THROW(gf1 / gf2, std::runtime_error);
    EXPECT_THROW(gf1 / 0, std::runtime_error);
    EXPECT_THROW(2 / gf2, std::runtime_error);

    EXPECT_THROW(gf1 /= 0, std::runtime_error);
    EXPECT_THROW(gf1 /= gf2, std::runtime_error);
}

TEST(GaloisField, Inversion) {
    GaloisField<PRIME> gf0{0};
    GaloisField<PRIME> gf1{1};
    GaloisField<PRIME> gf2{2};
    GaloisField<PRIME> gf3{3};
    GaloisField<PRIME> gf4{4};

    EXPECT_THROW(gf0.inverse(), std::runtime_error);
    ASSERT_EQ(gf1.inverse(), 1);
    ASSERT_EQ(gf2.inverse(), 3);
    ASSERT_EQ(gf3.inverse(), 2);
    ASSERT_EQ(gf4.inverse(), 4);
}

TEST(GaloisField, Comparison) {
    GaloisField<PRIME> gf1{3};
    GaloisField<PRIME> gf2{3};
    GaloisField<PRIME> gf3{4};
    GaloisField<PRIME> gf4{2};

    ASSERT_TRUE(gf1 == gf2);
    ASSERT_FALSE(gf3 == gf4);

    ASSERT_TRUE(gf1 == 3);
    ASSERT_FALSE(gf1 == 4);

    ASSERT_FALSE(gf1 != gf2);
    ASSERT_TRUE(gf3 != gf4);

    ASSERT_FALSE(gf1 != 3);
    ASSERT_TRUE(gf1 != 4);

    ASSERT_FALSE(gf1 < gf2);
    ASSERT_TRUE(gf1 < gf3);

    ASSERT_FALSE(gf1 < 3);
    ASSERT_TRUE(gf1 < 4);
    
    ASSERT_TRUE(gf1 > gf4);
    ASSERT_FALSE(gf1 > gf2);

    ASSERT_TRUE(gf1 > 2);
    ASSERT_FALSE(gf1 > 3);

    ASSERT_TRUE(gf1 <= gf2);
    ASSERT_FALSE(gf1 <= gf4);

    ASSERT_TRUE(gf1 <= 3);
    ASSERT_FALSE(gf1 <= 2);

    ASSERT_TRUE(gf1 >= gf2);
    ASSERT_FALSE(gf1 >= gf3);

    ASSERT_TRUE(gf1 >= 3);
    ASSERT_FALSE(gf1 >= 4);
}

int main() {
    ::testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}

File: lab2/java/src/test/java/com/galoisfield/GaloisFieldTest.java
package com.galoisfield;

import static org.junit.Assert.*;
import org.junit.Test;

public class GaloisFieldTest {
    @Test
    public void creatingCorrectObjects() {
        GaloisField gf = new GaloisField();

        assertEquals(GaloisField.P, gf.getCharacteristic());
        assertEquals(0, gf.toLong());

        GaloisField gf1 = new GaloisField(3);
        assertEquals(3, gf1.toLong());

        GaloisField gf2 = new GaloisField(7);
        assertEquals(2, gf2.toLong());

        GaloisField gf3 = new GaloisField(-16);
        assertEquals(4, gf3.toLong());
    }

    @Test
    public void addition() {
        GaloisField gf1 = new GaloisField(3);
        GaloisField gf2 = new GaloisField(4);

        assertEquals(2, GaloisField.add(gf1, gf2).toLong());
        assertEquals(1, GaloisField.add(gf2, 2).toLong());
        assertEquals(0, GaloisField.add(2, gf1).toLong());

        gf1.add(gf2);
        assertEquals(2, gf1.toLong());

        gf1.add(2);
        assertEquals(4, gf1.toLong());
    }

    @Test
    public void subtraction() {
        GaloisField gf1 = new GaloisField(3);
        GaloisField gf2 = new GaloisField(4);

        assertEquals(4, GaloisField.subtract(gf1, gf2).toLong());
        assertEquals(1, GaloisField.subtract(gf2, 3).toLong());
        assertEquals(1, GaloisField.subtract(4, gf1).toLong());

        gf1.subtract(gf2);
        assertEquals(4, gf1.toLong());

        gf1.subtract(2);
        assertEquals(2, gf1.toLong());
    }

    @Test
    public void multiplication() {
        GaloisField gf1 = new GaloisField(3);
        GaloisField gf2 = new GaloisField(4);

        assertEquals(2, GaloisField.multiply(gf1, gf2).toLong());
        assertEquals(2, GaloisField.multiply(gf2, 3).toLong());
        assertEquals(2, GaloisField.multiply(4, gf1).toLong());

        gf1.multiply(gf2);
        assertEquals(2, gf1.toLong());

        gf1.multiply(3);
        assertEquals(1, gf1.toLong());
    }

    @Test
    public void division() {
        GaloisField gf1 = new GaloisField(3);
        GaloisField gf2 = new GaloisField(4);

        assertEquals(2, GaloisField.divide(gf1, gf2).toLong());
        assertEquals(3, GaloisField.divide(gf2, 3).toLong());
        assertEquals(3, GaloisField.divide(4, gf1).toLong());

        gf1.divide(gf2);
        assertEquals(2, gf1.toLong());

        gf1.divide(3);
        assertEquals(4, gf1.toLong());
    }

    @Test
    public void inversion() {
        GaloisField gf1 = new GaloisField(3);

        assertEquals(2, GaloisField.inverse(gf1).toLong());
        assertEquals(4, GaloisField.inverse(4).toLong());

        gf1.inverse();
        assertEquals(2, gf1.toLong());
    }

    @Test
    public void equalism() {
        GaloisField a = new GaloisField(5);
        GaloisField b = new GaloisField(7);

        assertEquals(a, a);
        assertEquals(b, b);
        
        assertNotEquals(a, b);
        assertNotEquals(b, a);

        assertTrue(a.compareTo(a) == 0);
        assertTrue(b.compareTo(b) == 0);
        
        assertTrue(a.compareTo(b) < 0);
        assertTrue(b.compareTo(a) > 0);

        assertTrue(a.compareTo(5) == 0);
        assertTrue(b.compareTo(7) == 0);

        assertTrue(a.compareTo(7) < 0);
        assertTrue(b.compareTo(5) > 0);
    }

    @Test
    public void hashcode() {
        GaloisField a = new GaloisField(5);
        GaloisField b = new GaloisField(7);

        assertEquals(a.hashCode(), a.hashCode());
        assertNotEquals(a.hashCode(), b.hashCode());
    }
}
